### PR TITLE
service: nfc: Merge device interfaces and create the device manager

### PR DESCRIPTION
**Pull request by german77**
I always said nfc, mifare and amiibo are the same code. Well I got tired of editing 3 different places so in this PR I actually merged all implementations exactly as it should be from the beginning. This is an accuracy change more than anything for the end user everything will work the same.

### CODE REVIEW

File: src/core/hle/service/nfc/common/device_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+        devices[device_index] =
+            std::make_shared<NfcDevice>(Core::HID::IndexToNpadIdType(device_index), system,
+                                        service_context, availability_change_event);
+    }
+
+    is_initialized = false;
+}
+
+DeviceManager ::~DeviceManager() {
+    availability_change_event->Close();
```
Comment (liamwhite) [inline src/core/hle/service/nfc/common/device_manager.cpp:32]: Use service_context.CloseEvent, or close the readable handle too

File: src/core/hle/service/nfc/common/device_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+    return DeviceState::Unavailable;
+}
+
+Result DeviceManager::GetNpadId(u64 device_handle, Core::HID::NpadIdType& npad_id) const {
+    std::scoped_lock lock{mutex};
+
+    std::shared_ptr<NfcDevice> device = nullptr;
+    auto result = GetDeviceHandle(device_handle, device);
+
+    if (result.IsSuccess()) {
```
Comment (liamwhite) [inline src/core/hle/service/nfc/common/device_manager.cpp:88]: Optional, but I feel that this and all of the following methods would be greatly enhanced with the use of the R_TRY(), R_RETURN(), R_SUCCEED_IF(), R_SUCCEED() macros

File: src/core/hle/service/nfc/nfc_result.h [hunk 3]
```diff
@@ -1,23 +1,28 @@
...
 
-constexpr Result MifareDeviceNotFound(ErrorModule::NFCMifare, 64);
-constexpr Result MifareInvalidArgument(ErrorModule::NFCMifare, 65);
-constexpr Result MifareWrongDeviceState(ErrorModule::NFCMifare, 73);
-constexpr Result MifareNfcDisabled(ErrorModule::NFCMifare, 80);
-constexpr Result MifareTagRemoved(ErrorModule::NFCMifare, 97);
-constexpr Result MifareReadError(ErrorModule::NFCMifare, 288);
-
-} // namespace Service::NFC
+} // namespace Service::NFC
```
Comment (liamwhite) [inline src/core/hle/service/nfc/nfc_result.h:28]: Missing newline at end of file

File: src/core/hle/service/nfc/nfc_types.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    Type5, // ISO15693 RW/RO 540 bytes 106kbit/s
+};
+
+// This is nn::nfc::NfcProtocol
+// Verify this enum. It might be completely wrong default protocol is 0x48
+enum class NfcProtocol : u32 {
+    None,
+    TypeA = 1U << 0, // ISO14443A
+    TypeB = 1U << 1, // ISO14443B
+    TypeF = 1U << 2, // Sony Felica
```
Comment (liamwhite) [inline src/core/hle/service/nfc/nfc_types.h:61]: comment: name for these tags is FeliCa, right?

**Review by bunnei — APPROVED** [review]

**Pull request merged**